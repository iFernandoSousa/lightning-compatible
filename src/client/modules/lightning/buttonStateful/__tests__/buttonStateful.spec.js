/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

import { createElement } from 'lwc';
import Element from 'c/buttonStateful';

const createButtonStateful = attributes => {
    const element = createElement('c-button-stateful', { is: Element });
    Object.assign(element, attributes);
    document.body.appendChild(element);

    return element;
};

describe('c-button-stateful', () => {
    it('default', () => {
        const element = createButtonStateful({
            labelWhenOff: 'Follow',
            labelWhenOn: 'Following',
            labelWhenHover: 'Unfollow'
        });

        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
        });
    });

    describe('testIconNames', () => {
        it('Expected icons should render for each state', () => {
            const element = createButtonStateful({
                iconNameWhenOff: 'utility:add',
                iconNameWhenOn: 'utility:check',
                iconNameWhenHover: 'utility:close',
                labelWhenOff: 'Follow',
                labelWhenOn: 'Following',
                labelWhenHover: 'Unfollow'
            });

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('Invalid or empty icon names should render empty icon elements', () => {
            const element = createButtonStateful({
                iconNameWhenOff: 'utility:',
                iconNameWhenOn: '',
                iconNameWhenHover: 'utility:pickle',
                labelWhenOff: 'Follow',
                labelWhenOn: 'Following',
                labelWhenHover: 'Unfollow'
            });

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });
    });

    describe('Variant', () => {
        const createButtonStatefulWithVariant = variant => {
            const element = createButtonStateful({
                iconNameWhenOff: 'utility:add',
                iconNameWhenOn: 'utility:check',
                iconNameWhenHover: 'utility:close',
                labelWhenOff: 'Follow',
                labelWhenOn: 'Following',
                labelWhenHover: 'Unfollow',
                variant
            });

            return element;
        };

        it('should add default classes on invalid variant', () => {
            const element = createButtonStatefulWithVariant('invalid');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should add correct classes on neutral variant', () => {
            const element = createButtonStatefulWithVariant('neutral');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should add correct classes on brand variant', () => {
            const element = createButtonStatefulWithVariant('brand');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should add correct classes on inverse variant', () => {
            const element = createButtonStatefulWithVariant('inverse');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should add correct classes on text variant', () => {
            const element = createButtonStatefulWithVariant('text');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });
    });

    describe('testLabels', () => {
        it('Expected labels should render for each state', () => {
            const element = createButtonStateful({
                labelWhenOff: 'Seek',
                labelWhenOn: 'Seeking',
                labelWhenHover: 'Ignore'
            });

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('`labelWhenOn` value should be used when `labelWhenHover` is missing or invalid', () => {
            const element = createButtonStateful({
                labelWhenOff: 'Seek',
                labelWhenOn: 'Seeking'
            });

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });
    });
});
