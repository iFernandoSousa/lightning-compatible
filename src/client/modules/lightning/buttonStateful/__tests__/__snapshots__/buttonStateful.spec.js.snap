// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-button-stateful Variant should add correct classes on brand variant 1`] = `
<c-button-stateful>
  #shadow-root(open)
    <button
      aria-live="assertive"
      class="slds-button slds-button_stateful slds-button_brand slds-not-selected"
    >
      <span
        class="slds-text-not-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="add"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#add"
              />
            </svg>
        </c-primitive-icon>
        Follow
      </span>
      <span
        class="slds-text-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="check"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#check"
              />
            </svg>
        </c-primitive-icon>
        Following
      </span>
      <span
        class="slds-text-selected-focus"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="close"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#close"
              />
            </svg>
        </c-primitive-icon>
        Unfollow
      </span>
    </button>
</c-button-stateful>
`;

exports[`c-button-stateful Variant should add correct classes on inverse variant 1`] = `
<c-button-stateful>
  #shadow-root(open)
    <button
      aria-live="assertive"
      class="slds-button slds-button_stateful slds-button_inverse slds-not-selected"
    >
      <span
        class="slds-text-not-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="add"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#add"
              />
            </svg>
        </c-primitive-icon>
        Follow
      </span>
      <span
        class="slds-text-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="check"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#check"
              />
            </svg>
        </c-primitive-icon>
        Following
      </span>
      <span
        class="slds-text-selected-focus"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="close"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#close"
              />
            </svg>
        </c-primitive-icon>
        Unfollow
      </span>
    </button>
</c-button-stateful>
`;

exports[`c-button-stateful Variant should add correct classes on neutral variant 1`] = `
<c-button-stateful>
  #shadow-root(open)
    <button
      aria-live="assertive"
      class="slds-button slds-button_stateful slds-button_neutral slds-not-selected"
    >
      <span
        class="slds-text-not-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="add"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#add"
              />
            </svg>
        </c-primitive-icon>
        Follow
      </span>
      <span
        class="slds-text-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="check"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#check"
              />
            </svg>
        </c-primitive-icon>
        Following
      </span>
      <span
        class="slds-text-selected-focus"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="close"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#close"
              />
            </svg>
        </c-primitive-icon>
        Unfollow
      </span>
    </button>
</c-button-stateful>
`;

exports[`c-button-stateful Variant should add correct classes on text variant 1`] = `
<c-button-stateful>
  #shadow-root(open)
    <button
      aria-live="assertive"
      class="slds-button slds-button_stateful slds-not-selected"
    >
      <span
        class="slds-text-not-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="add"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#add"
              />
            </svg>
        </c-primitive-icon>
        Follow
      </span>
      <span
        class="slds-text-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="check"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#check"
              />
            </svg>
        </c-primitive-icon>
        Following
      </span>
      <span
        class="slds-text-selected-focus"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="close"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#close"
              />
            </svg>
        </c-primitive-icon>
        Unfollow
      </span>
    </button>
</c-button-stateful>
`;

exports[`c-button-stateful Variant should add default classes on invalid variant 1`] = `
<c-button-stateful>
  #shadow-root(open)
    <button
      aria-live="assertive"
      class="slds-button slds-button_stateful slds-button_neutral slds-not-selected"
    >
      <span
        class="slds-text-not-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="add"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#add"
              />
            </svg>
        </c-primitive-icon>
        Follow
      </span>
      <span
        class="slds-text-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="check"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#check"
              />
            </svg>
        </c-primitive-icon>
        Following
      </span>
      <span
        class="slds-text-selected-focus"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="close"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#close"
              />
            </svg>
        </c-primitive-icon>
        Unfollow
      </span>
    </button>
</c-button-stateful>
`;

exports[`c-button-stateful default 1`] = `
<c-button-stateful>
  #shadow-root(open)
    <button
      aria-live="assertive"
      class="slds-button slds-button_stateful slds-button_neutral slds-not-selected"
    >
      <span
        class="slds-text-not-selected"
      >
        Follow
      </span>
      <span
        class="slds-text-selected"
      >
        Following
      </span>
      <span
        class="slds-text-selected-focus"
      >
        Unfollow
      </span>
    </button>
</c-button-stateful>
`;

exports[`c-button-stateful testIconNames Expected icons should render for each state 1`] = `
<c-button-stateful>
  #shadow-root(open)
    <button
      aria-live="assertive"
      class="slds-button slds-button_stateful slds-button_neutral slds-not-selected"
    >
      <span
        class="slds-text-not-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="add"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#add"
              />
            </svg>
        </c-primitive-icon>
        Follow
      </span>
      <span
        class="slds-text-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="check"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#check"
              />
            </svg>
        </c-primitive-icon>
        Following
      </span>
      <span
        class="slds-text-selected-focus"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="close"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#close"
              />
            </svg>
        </c-primitive-icon>
        Unfollow
      </span>
    </button>
</c-button-stateful>
`;

exports[`c-button-stateful testIconNames Invalid or empty icon names should render empty icon elements 1`] = `
<c-button-stateful>
  #shadow-root(open)
    <button
      aria-live="assertive"
      class="slds-button slds-button_stateful slds-button_neutral slds-not-selected"
    >
      <span
        class="slds-text-not-selected"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key=""
              focusable="false"
            >
              <use
                xlink:href=""
              />
            </svg>
        </c-primitive-icon>
        Follow
      </span>
      <span
        class="slds-text-selected"
      >
        Following
      </span>
      <span
        class="slds-text-selected-focus"
      >
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-button__icon_stateful slds-button__icon_left"
              data-key="pickle"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#pickle"
              />
            </svg>
        </c-primitive-icon>
        Unfollow
      </span>
    </button>
</c-button-stateful>
`;

exports[`c-button-stateful testLabels \`labelWhenOn\` value should be used when \`labelWhenHover\` is missing or invalid 1`] = `
<c-button-stateful>
  #shadow-root(open)
    <button
      aria-live="assertive"
      class="slds-button slds-button_stateful slds-button_neutral slds-not-selected"
    >
      <span
        class="slds-text-not-selected"
      >
        Seek
      </span>
      <span
        class="slds-text-selected"
      >
        Seeking
      </span>
      <span
        class="slds-text-selected-focus"
      >
        Seeking
      </span>
    </button>
</c-button-stateful>
`;

exports[`c-button-stateful testLabels Expected labels should render for each state 1`] = `
<c-button-stateful>
  #shadow-root(open)
    <button
      aria-live="assertive"
      class="slds-button slds-button_stateful slds-button_neutral slds-not-selected"
    >
      <span
        class="slds-text-not-selected"
      >
        Seek
      </span>
      <span
        class="slds-text-selected"
      >
        Seeking
      </span>
      <span
        class="slds-text-selected-focus"
      >
        Ignore
      </span>
    </button>
</c-button-stateful>
`;
